mongo/mongo.go

```diff
@@ -421,26 +430,7 @@ func marshalValue(
 	bsonOpts *options.BSONOptions,
 	registry *bsoncodec.Registry,
 ) (bsoncore.Value, error) {
-	if registry == nil {
-		registry = bson.DefaultRegistry
-	}
-	if val == nil {
-		return bsoncore.Value{}, ErrNilValue
-	}
-
-	buf := new(bytes.Buffer)
-	enc, err := getEncoder(buf, bsonOpts, registry)
-	if err != nil {
-		return bsoncore.Value{}, fmt.Errorf("error configuring BSON encoder: %w", err)
-	}
-
-	// Encode the value in a single-element document with an empty key. Use bsoncore to extract the
-	// first element and return the BSON value.
-	err = enc.Encode(bson.D{{Key: "", Value: val}})
-	if err != nil {
-		return bsoncore.Value{}, MarshalError{Value: val, Err: err}
-	}
-	return bsoncore.Document(buf.Bytes()).Index(0).Value(), nil
+	return codecutil.MarshalValue(val, registry, newEncoderFn(bsonOpts))
```

Comment:
Optional: Now that the logic is moved to `codecutil`, this `marshalValue` wrapper function is unnecessary. Consider removing it and replacing all calls to it with `codecutil.MarshalValue` instead.

Reply:
The "marshalValue" function is used ~30 times in pretty differing cases, so this would be quite a large change. Also, I think ultimately the codecutil.MarshalValue function will eventually be deprecated if/when the x package BatchCursors is merged into the mongo package. In this case, we would return the logic back to marshalValue.